pkg/util/shared_action_queue.go

```diff
@@ -0,0 +1,100 @@
+package util
+
+import (
+	"context"
+	"sync"
+)
+
+type ActionContext struct {
+	ctx        context.Context
+	cancelFunc context.CancelFunc
+	wg         sync.WaitGroup
+	results    chan error
+	isWaiting  bool
+}
+
+type Action = func(*ActionContext)
+type ActionGroup = []Action
+
+type SharedActionQueue struct {
+	sharedActionQueue chan func()
+}
+
+// This is a utility designed to run 'Action's to do some work async but limit
+// by a pre-defined number. This allows multiple components to share the same
+// pool of goroutines or other similar resources (like data streams) easily.
+func NewSharedActionQueue(maxConcurrentDownloads uint32) *SharedActionQueue {
```

Comment:
`s/maxConcurrentDownloads/maxConcurrency/`? This code is quite independent from doing downloads.